File: web/src/types/commission.ts
// Tipos de Comissão e Estorno — espelha backend/internal/dto/commission.go
// Task 7.1.3: Commission, CommissionReversal, ApurationRequest, ApurationResult, CommissionNetBalance
// Ref: contracts/api.md §/commissions; spec §FR-020, FR-027-029; P-III (sem float)

export type CommissionStatus = 'pendente' | 'aprovado' | 'pago';
export type ReversalStatus = 'aplicado' | 'pendente_aprovacao' | 'aprovado' | 'lancado';

export interface Commission {
  id: string; // UUID
  orderId: string;
  vendorId: string;
  valueCents: number; // int64 centavos (P-III — nunca float)
  netCents: number;   // líquido após estornos (P-III)
  appliedPercentage: string; // "5.5000" — string para preservar precisão (P-III)
  ruleId: string;
  periodYear: number;
  periodMonth: number; // 1-12
  status: CommissionStatus;
  calculatedAt: string; // ISO 8601 UTC
}

export interface CommissionReversal {
  id: string; // UUID
  commissionId: string;
  orderId: string;
  valueCents: number; // negativo (≤ 0) — estorno reduz saldo (P-III)
  status: ReversalStatus;
  createdAt: string; // ISO 8601 UTC
  actorUserId: string; // auditabilidade P-I
}

export interface CommissionNetBalance {
  commissionId: string;
  vendorId: string;
  netCents: number; // saldo líquido = value_cents + SUM(reversals.value_cents) (P-III)
}

// ApurationRequest — corpo para POST /commissions/apurate
export interface ApurationRequest {
  year: number;
  month: number; // 1-12
}

// ApurationResult — resposta de POST /commissions/apurate
export interface ApurationResult {
  periodYear: number;
  periodMonth: number;
  calculated: number; // comissões calculadas/inseridas
  skipped: number;    // comissões puladas por idempotência (FR-014)
  totalCents: number; // soma das comissões inseridas (P-III — nunca float)
}

// TransitionCommissionRequest — corpo para PATCH /commissions/{id}/status
export interface TransitionCommissionRequest {
  status: CommissionStatus;
  motivo?: string; // obrigatório ao reverter aprovado→pendente
}

File: backend/internal/dto/commission.go
// Package dto — Commission DTOs com json tags camelCase para a API REST.
// Constitution P-III: valores monetários como int64 (centavos); percentuais como string.
// Constitution P-I: todos os campos de auditoria são expostos (calculatedAt, status).
package dto

import (
	"financial-dashboard/backend/internal/repository"
)

// CommissionResponse é o DTO de saída para uma comissão.
type CommissionResponse struct {
	ID                string `json:"id"`
	OrderID           string `json:"orderId"`
	VendorID          string `json:"vendorId"`
	ValueCents        int64  `json:"valueCents"`        // bruto em centavos (P-III: int64)
	NetCents          int64  `json:"netCents"`          // líquido após estornos (P-III: int64)
	AppliedPercentage string `json:"appliedPercentage"` // "8.0000" — string para preservar precisão
	RuleID            string `json:"ruleId"`
	PeriodYear        int    `json:"periodYear"`
	PeriodMonth       int    `json:"periodMonth"`
	Status            string `json:"status"`
	CalculatedAt      string `json:"calculatedAt"`
}

// CommissionFromDomain converte um repository.Commission para CommissionResponse.
func CommissionFromDomain(c repository.Commission) CommissionResponse {
	return CommissionResponse{
		ID:                c.ID,
		OrderID:           c.OrderID,
		VendorID:          c.VendorID,
		ValueCents:        c.ValueCents,
		NetCents:          c.NetCents,
		AppliedPercentage: c.AppliedPercentage.String(),
		RuleID:            c.RuleID,
		PeriodYear:        c.PeriodYear,
		PeriodMonth:       c.PeriodMonth,
		Status:            string(c.Status),
		CalculatedAt:      c.CalculatedAt.UTC().Format("2006-01-02T15:04:05Z"),
	}
}

// ApurationResultResponse é o DTO de saída para o resultado de apuração.
type ApurationResultResponse struct {
	PeriodYear  int   `json:"periodYear"`
	PeriodMonth int   `json:"periodMonth"`
	Calculated  int   `json:"calculated"`   // número de comissões calculadas/inseridas
	Skipped     int   `json:"skipped"`      // número de comissões puladas (idempotência)
	TotalCents  int64 `json:"totalCents"`   // soma das comissões inseridas em centavos (P-III: int64)
}

// TransitionCommissionRequest é o DTO de entrada para transição de status.
type TransitionCommissionRequest struct {
	Status string `json:"status"` // "aprovado", "pago" ou "pendente"
	Motivo string `json:"motivo"` // obrigatório para aprovado→pendente
}

// ApurateRequest é o DTO de entrada para apuração mensal.
type ApurateRequest struct {
	Year  int `json:"year"`
	Month int `json:"month"`
}
